File: kwi-ui/src/lib/commands.ts
import { invoke } from "@tauri-apps/api/core";
import type { Project, Area, WorkItem, RelatedItem } from "./types";

export async function checkConnection(): Promise<void> {
  return invoke("check_connection");
}

export async function listProjects(): Promise<Project[]> {
  return invoke("list_projects");
}

export async function listAreas(projectId: number): Promise<Area[]> {
  return invoke("list_areas", { projectId });
}

export async function listWorkItems(
  projectId?: number,
  areaId?: number,
  wiType?: string,
  wiStatus?: string,
  showArchived?: boolean,
): Promise<WorkItem[]> {
  return invoke("list_work_items", {
    projectId,
    areaId,
    wiType,
    wiStatus,
    showArchived,
  });
}

export async function getWorkItem(id: number): Promise<WorkItem> {
  return invoke("get_work_item", { id });
}

export async function getValidTypes(): Promise<string[]> {
  return invoke("get_valid_types");
}

export async function getValidStatuses(): Promise<string[]> {
  return invoke("get_valid_statuses");
}

export async function getValidTshirtSizes(): Promise<string[]> {
  return invoke("get_valid_tshirt_sizes");
}

export async function createWorkItem(params: {
  projectId: number;
  title: string;
  content: string;
  wiType: string;
  wiStatus?: string;
  wiTshirt?: string;
  areaId?: number;
  sprint?: string;
  details?: string;
  parentId?: number;
}): Promise<WorkItem> {
  return invoke("create_work_item", params);
}

export async function updateWorkItem(params: {
  id: number;
  title?: string;
  content?: string;
  wiType?: string;
  wiStatus?: string;
  wiTshirt?: string;
  areaId?: number;
  sprint?: string;
  details?: string;
  parentId?: number;
}): Promise<WorkItem> {
  return invoke("update_work_item", params);
}

export async function archiveWorkItem(id: number): Promise<WorkItem> {
  return invoke("archive_work_item", { id });
}

export async function searchWorkItems(
  query: string,
  projectId?: number,
): Promise<WorkItem[]> {
  return invoke("search_work_items", { query, projectId });
}

export async function listRelated(workItemId: number): Promise<RelatedItem[]> {
  return invoke("list_related", { workItemId });
}

export async function relateWorkItems(
  leftId: number,
  rightId: number,
  relationship: string,
): Promise<void> {
  return invoke("relate_work_items", { leftId, rightId, relationship });
}

export async function unrelateWorkItems(
  leftId: number,
  rightId: number,
): Promise<void> {
  return invoke("unrelate_work_items", { leftId, rightId });
}

export async function createProject(params: {
  project: string;
  cnPath: string;
  ghRepo?: string;
  description?: string;
}): Promise<Project> {
  return invoke("create_project", params);
}

export async function createArea(params: {
  projectId: number;
  name: string;
  description?: string;
}): Promise<Area> {
  return invoke("create_area", params);
}

File: kwi-ui/src-tauri/src/commands.rs
use deadpool_postgres::Pool;
use tauri::State;

use crate::models::{Area, Project, RelatedItem, WorkItem};
use crate::queries;

pub struct AppState {
    pub pool: Option<Pool>,
    pub connection_error: Option<String>,
}

fn pool_or_err(state: &AppState) -> Result<&Pool, String> {
    state.pool.as_ref().ok_or_else(|| {
        state
            .connection_error
            .clone()
            .unwrap_or_else(|| "Database not configured".to_string())
    })
}

#[tauri::command]
pub async fn check_connection(state: State<'_, AppState>) -> Result<(), String> {
    pool_or_err(&state)?;
    Ok(())
}

#[tauri::command]
pub async fn list_projects(state: State<'_, AppState>) -> Result<Vec<Project>, String> {
    queries::list_projects(pool_or_err(&state)?).await
}

#[tauri::command]
pub async fn list_areas(state: State<'_, AppState>, project_id: i32) -> Result<Vec<Area>, String> {
    queries::list_areas(pool_or_err(&state)?, project_id).await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn list_work_items(
    state: State<'_, AppState>,
    project_id: Option<i32>,
    area_id: Option<i32>,
    wi_type: Option<String>,
    wi_status: Option<String>,
    show_archived: Option<bool>,
) -> Result<Vec<WorkItem>, String> {
    queries::list_work_items(
        pool_or_err(&state)?,
        project_id,
        area_id,
        wi_type,
        wi_status,
        show_archived,
    )
    .await
}

#[tauri::command]
pub async fn get_work_item(state: State<'_, AppState>, id: i32) -> Result<WorkItem, String> {
    queries::get_work_item(pool_or_err(&state)?, id).await
}

#[tauri::command]
pub async fn get_valid_types(state: State<'_, AppState>) -> Result<Vec<String>, String> {
    queries::get_valid_types(pool_or_err(&state)?).await
}

#[tauri::command]
pub async fn get_valid_statuses(state: State<'_, AppState>) -> Result<Vec<String>, String> {
    queries::get_valid_statuses(pool_or_err(&state)?).await
}

#[tauri::command]
pub async fn get_valid_tshirt_sizes() -> Result<Vec<String>, String> {
    Ok(queries::get_valid_tshirt_sizes())
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn create_work_item(
    state: State<'_, AppState>,
    project_id: i32,
    title: String,
    content: String,
    wi_type: String,
    wi_status: Option<String>,
    wi_tshirt: Option<String>,
    area_id: Option<i32>,
    sprint: Option<String>,
    details: Option<String>,
    parent_id: Option<i32>,
) -> Result<WorkItem, String> {
    queries::create_work_item(
        pool_or_err(&state)?,
        project_id,
        title,
        content,
        wi_type,
        wi_status,
        wi_tshirt,
        area_id,
        sprint,
        details,
        parent_id,
    )
    .await
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn update_work_item(
    state: State<'_, AppState>,
    id: i32,
    title: Option<String>,
    content: Option<String>,
    wi_type: Option<String>,
    wi_status: Option<String>,
    wi_tshirt: Option<String>,
    area_id: Option<i32>,
    sprint: Option<String>,
    details: Option<String>,
    parent_id: Option<i32>,
) -> Result<WorkItem, String> {
    queries::update_work_item(
        pool_or_err(&state)?,
        id,
        title,
        content,
        wi_type,
        wi_status,
        wi_tshirt,
        area_id,
        sprint,
        details,
        parent_id,
    )
    .await
}

#[tauri::command]
pub async fn archive_work_item(state: State<'_, AppState>, id: i32) -> Result<WorkItem, String> {
    queries::archive_work_item(pool_or_err(&state)?, id).await
}

#[tauri::command]
pub async fn unarchive_work_item(state: State<'_, AppState>, id: i32) -> Result<WorkItem, String> {
    queries::unarchive_work_item(pool_or_err(&state)?, id).await
}

#[tauri::command]
pub async fn search_work_items(
    state: State<'_, AppState>,
    query: String,
    project_id: Option<i32>,
) -> Result<Vec<WorkItem>, String> {
    queries::search_work_items(pool_or_err(&state)?, query, project_id).await
}

#[tauri::command]
pub async fn list_related(
    state: State<'_, AppState>,
    work_item_id: i32,
) -> Result<Vec<RelatedItem>, String> {
    queries::list_related(pool_or_err(&state)?, work_item_id).await
}

#[tauri::command]
pub async fn relate_work_items(
    state: State<'_, AppState>,
    left_id: i32,
    right_id: i32,
    relationship: String,
) -> Result<(), String> {
    queries::relate_work_items(pool_or_err(&state)?, left_id, right_id, relationship).await
}

#[tauri::command]
pub async fn unrelate_work_items(
    state: State<'_, AppState>,
    left_id: i32,
    right_id: i32,
) -> Result<(), String> {
    queries::unrelate_work_items(pool_or_err(&state)?, left_id, right_id).await
}

#[tauri::command]
pub async fn create_project(
    state: State<'_, AppState>,
    project: String,
    cn_path: String,
    gh_repo: Option<String>,
    description: Option<String>,
) -> Result<Project, String> {
    queries::create_project(pool_or_err(&state)?, project, cn_path, gh_repo, description).await
}

#[tauri::command]
pub async fn create_area(
    state: State<'_, AppState>,
    project_id: i32,
    name: String,
    description: Option<String>,
) -> Result<Area, String> {
    queries::create_area(pool_or_err(&state)?, project_id, name, description).await
}
